### PR TITLE
Pr enhance argreconcile 245

### LINKED ISSUE #245: DRY: factoring out the argument cleanup that many rio.import_* and rio.export_* functions need.
Please specify whether your issue is about:

 - [ ] a possible bug
 - [ ] a question about package functionality
 - [x] a suggested code or documentation change, improvement to the code, or feature request

I would like to volunteer to fix #241 (actually I meant to submit a PR after I posted it, but got busy with a few other things). Looking over it I see an opportunity to simplify the individual import and export functions and make them more maintainable. Namely generalizing the `args <- args[intersect(args,formals(fn))]` pattern in the function sketched out below (not yet thoroughly tested, just proof of concept). While we're at it, we could include additional safeguards: removing duplicate arguments and unnamed arguments, and re-mapping certain argument names. This function could safely be run on all underlying functions-- if they don't need argument filtering because they have `...` among their formals, they would still benefit from remapping and protection against duplication (to which `...` functions are particularly vulnerable otherwise).

I am _not_ proposing to cut over all the functions at once. I am proposing to use this for fixing current and future unknown-argument bugs as they are noticed. Possibly updating and testing and already patched functions such as in #221 and #223 individually as time permits.

I understand reluctance to introduce an upstream dependency on `R.utils::DoCall` which has similar functionality. The below code only uses base R.

So, if this is acceptable, I would like to submit this as a PR, and follow that one with a PR that actually closes #241 

Put your code here:

```R
#' Adapt an argument list to a function excluding arguments that will not be 
#' recognized by it, redundant arguments, and un-named arguments.
#'
#' @param fn A function to which an argument list needs to be adapted.
#' @param ... An arbitrary list of named arguments (unnamed ones will be ignored)
#' @param .args A list or \code{alist} of named arguments, to be merged with \code{...}
#' @param .docall If set to \code{TRUE} will not only clean up the arguments but 
#'                also execute \code{fn} with those arguments (\code{FALSE} by
#'                default)
#' @param .remap An optional named character vector or named list of character 
#'               values for standardizing arguments that play the same role but
#'               have different names in different functions.
#' @param .warn Whether to issue a warning message (default) when invalid 
#'              arguments need to be discarded.
#'
#' @return Either a named list or the result of calling \code{fn} with the 
#'         supplied arguments
#'
#' @examples
#' 
#' arg_reconcile(p.adjust,bla='aaa',baz='xzzz',n=40,q=c(.1,.2,.02,.3,1,0,1),.remap = c(q='p'))
#' arg_reconcile(import,file='hello.csv',file='goodbye.csv',
#'               bla='aaa',baz='xzzz',format='csv')
arg_reconcile <- function(fn,...,.args=alist(),.docall=FALSE,
                          .remap=list(),
                          .warn=TRUE){
  # capture the formal arguments of the target function
  frmls <- formals(fn)
  # capture the arguments, both freeform and an explicit list
  args <- c(list(...),.args)
  # get rid of duplicate arguments
  for(ii in names(args)) {
    if( sum(dupe <- names(args) %in% ii)>1) {
      args[seq_along(args)[dupe][-which(TRUE,dupe)]] <- NULL
    }
  }
  # if any remappings of one argument to another are specified, perform them
  for( ii in names(.remap) ){
    if( !.remap[[ii]] %in% names(args) && ii %in% names(args) ){
      args[[.remap[[ii]] ]] <- args[[ii]]}
  }
  # remove any unnamed arguments
  args[names(args)==""] <- NULL
  # if the target function doesn't have '...' as an argument, check to make sure
  # only recognized arguments get passed, optionally with a warning
  if ( !'...' %in% names(frmls) ){
    unused <- setdiff(names(args),names(frmls))
    if ( length(unused)>0 && .warn ){
      warning("The following arguments were ignored for ",substitute(fn),":\n",
              paste(unused, collapse = ', '))
      args <- args[intersect(names(args),names(frmls))]
    }
  }
  # the final, cleaned-up arguments either get used on the function or returned
  # as a list, depending on how .docall is set
  if ( .docall ) return(do.call(fn,args)) else return(args);
}

```

### PR DESCRIPTION
**Pull request by bokov**
Please ensure the following before submitting a PR:

 - [x] if suggesting code changes or improvements, [open an issue](https://github.com/leeper/rio/issues/new) first 
      closes #245 and maybe #231 
 - [x] for all but trivial changes (e.g., typo fixes), add your name to [DESCRIPTION](https://github.com/leeper/rio/blob/master/DESCRIPTION)
 - [x] for all but trivial changes (e.g., typo fixes), documentation your change in [NEWS.md](https://github.com/leeper/rio/blob/master/NEWS.md) with a parenthetical reference to the issue number being addressed
 - [x] if changing documentation, edit files in `/R` not `/man` and run `devtools::document()` to update documentation
 - [x] add code or new test files to [`/tests`](https://github.com/leeper/rio/tree/master/tests/testthat) for any new functionality or bug fix
 - [x] make sure `R CMD check` runs without error before submitting the PR

An arg_reconcile() function to generically remap and de-duplicate arguments for the function specified as its first argument.

### CODE REVIEW

File: R/arg_reconcile.R [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+#'              arguments. 
+#'
+#' @param fn      A function to which an argument list needs to be adapted. Use
+#'                the unquoted name of the function. If it's in a different 
+#'                package then the fully qualified unquoted name (e.g. 
+#'                \code{utils::read.table})
+#' @param ...     An arbitrary list of named arguments (unnamed ones will be 
+#'                ignored). Arguments in \code{.args} are overridden by
+#'                arguments of the same name (if any) in \code{...}
+#' @param .args   A list or \code{alist} of named arguments, to be merged
```
Comment (bokov) [inline R/arg_reconcile.R]: The rest of these argument names all begin with `.`  to reduce likelihood of collision with `... ` but I could use some other prefix if needed by project style conventions

File: R/arg_reconcile.R [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+#'         supplied arguments
+#'
+#' @examples
+#' 
+#' arg_reconcile(p.adjust, bla = "aaa", baz = "xzzz", n = 40, 
+#'                     q = c(.1, .2, .02, .3, 1, 0, 1), .remap = c(q = "p"))
+#' arg_reconcile(import, file = "hello.csv", file = "goodbye.csv", bla = "aaa", 
+#'               baz = "xzzz", format = "csv")
+#' 
+#' @export
```
Comment (leeper) [inline R/arg_reconcile.R]: Main reaction at this point is that I don't think we need to export this function.

Comment (bokov) [reply R/arg_reconcile.R]: Done!

**Comment by bokov** [comment]
Actually, can you wait a little before accepting this PR? I found a bug that I need to fix.

**Comment by bokov** [comment]
Okay, bug fixed, and PR is ready for review again.

**Comment by bokov** [comment]
Hello, I'm just following up to see if there are any further changes I should make to this PR. Thanks.

File: R/arg_reconcile.R [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+#'         supplied arguments
+#'
+arg_reconcile <- function(fn, ..., .args = alist(), .docall = FALSE, 
+                          .wtlist = c(), .bklist= c(), .remap = list(), 
+                          .warn = TRUE, .error = "default", .finish = identity) {
+  # capture the formal arguments of the target function
+  frmls <- formals(fn)
+  # both freeform and an explicit list
+  args <- match.call(expand.dots = FALSE)$`...`
+  if (.docall) {
```
Comment (leeper) [inline R/arg_reconcile.R:56]: Can we be consistent with the rest of the package on brackets/formatting - always use {} even on short expressions; else always on a separate line; no spaces after `(` or before `)` in condition statements. Always like:

```
if (cond) {
  # stuff
} else { 
  # other stuff
}

for (i in seq) {
  # stuff
}
```

File: R/arg_reconcile.R [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+#' @return Either a named list or the result of calling \code{fn} with the 
+#'         supplied arguments
+#'
+arg_reconcile <- function(fn, ..., .args = alist(), .docall = FALSE, 
+                          .wtlist = c(), .bklist= c(), .remap = list(), 
+                          .warn = TRUE, .error = "default", .finish = identity) {
+  # capture the formal arguments of the target function
+  frmls <- formals(fn)
+  # both freeform and an explicit list
+  args <- match.call(expand.dots = FALSE)$`...`
```
Comment (leeper) [inline R/arg_reconcile.R:55]: Don't use $ notation:

`args <- match.call(expand.dots = FALSE)[["..."]]`

File: R/arg_reconcile.R [hunk 5]
```diff
@@ -0,0 +1,115 @@
+#' @title Reconcile an argument list to any function signature.
+#' 
+#' @description Adapt an argument list to a function excluding arguments that 
+#'              will not be recognized by it, redundant arguments, and un-named
+#'              arguments. 
+#'
+#' @param fn      A function to which an argument list needs to be adapted. Use
```
Comment (leeper) [inline R/arg_reconcile.R:7]: `fun` instead of `fn`

File: R/arg_reconcile.R [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+#' @param ...     An arbitrary list of named arguments (unnamed ones will be 
+#'                ignored). Arguments in \code{.args} are overridden by
+#'                arguments of the same name (if any) in \code{...}
+#' @param .args   A list or \code{alist} of named arguments, to be merged 
+#'                with \code{...}. Arguments in \code{.args} are overridden by
+#'                arguments of the same name (if any) in \code{...}
+#' @param .docall If set to \code{TRUE} will not only clean up the arguments but 
+#'                also execute \code{fn} with those arguments (\code{FALSE} by
+#'                default) and return the results
+#' @param .wtlist Whitelist. If not empty, only arguments named here will be
```
Comment (leeper) [inline R/arg_reconcile.R:20]: `.include` instead of `.wtlist`

File: R/arg_reconcile.R [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+#' @param .args   A list or \code{alist} of named arguments, to be merged 
+#'                with \code{...}. Arguments in \code{.args} are overridden by
+#'                arguments of the same name (if any) in \code{...}
+#' @param .docall If set to \code{TRUE} will not only clean up the arguments but 
+#'                also execute \code{fn} with those arguments (\code{FALSE} by
+#'                default) and return the results
+#' @param .wtlist Whitelist. If not empty, only arguments named here will be 
+#'                permitted, and only if they satisfy the conditions implied by
+#'                the other arguments. Evaluated before \code{.remap}.
+#' @param .bklist Blacklist. If not empty, arguments named here will be removed
```
Comment (leeper) [inline R/arg_reconcile.R:23]: `.exclude` instead of `.bklist`

File: R/arg_reconcile.R [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+                          .warn = TRUE, .error = "default", .finish = identity) {
+  # capture the formal arguments of the target function
+  frmls <- formals(fn)
+  # both freeform and an explicit list
+  args <- match.call(expand.dots = FALSE)$`...`
+  if (.docall) {
+    for (ii in names(args)) try(args[[ii]] <- eval(args[[ii]], parent.frame()))
+  }
+  # get rid of duplicate arguments, with freeform arguments 
+  dupes <- table(names(args))
```
Comment (leeper) [inline R/arg_reconcile.R:60]: Can't you replace these two lines with `duplicated(names(args))` ?

File: R/arg_reconcile.R [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+    }
+  }
+  # remove any unnamed arguments
+  args[names(args) == ""] <- NULL
+  # if the target function doesn't have "..." as an argument, check to make sure
+  # only recognized arguments get passed, optionally with a warning
+  if (!"..." %in% names(frmls) ) {
+    unused <- setdiff(names(args), names(frmls))
+    if ( length(unused)>0 ){
+      if (.warn) {
```
Comment (leeper) [inline R/arg_reconcile.R:87]: `if (isTRUE(.warn)) {`

File: R/arg_reconcile.R [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+      args[[.remap[[ii]] ]] <- args[[ii]]
+    }
+  }
+  # remove any unnamed arguments
+  args[names(args) == ""] <- NULL
+  # if the target function doesn't have "..." as an argument, check to make sure
+  # only recognized arguments get passed, optionally with a warning
+  if (!"..." %in% names(frmls) ) {
+    unused <- setdiff(names(args), names(frmls))
+    if ( length(unused)>0 ){
```
Comment (leeper) [inline R/arg_reconcile.R:86]: Formatting: `if (length(unused) > 0) {`

File: R/arg_reconcile.R [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+      if (.warn) {
+        warning("The following arguments were ignored for ", 
+                deparse(substitute(fn)), ":\n", paste(unused, collapse = ", "))
+      }
+      args <- args[intersect(names(args), names(frmls))]
+    }
+  }
+  # the final, cleaned-up arguments either get returned as a list or used on the 
+  # function, depending on how .docall is set
+  if ( !.docall ) return(args) else {
```
Comment (leeper) [inline R/arg_reconcile.R:96]: add {} and else on separate line

File: R/arg_reconcile.R [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+    # run the function and return the result case
+    oo <- try(do.call(fn, args),silent = TRUE)
+    if (!methods::is(oo, "try-error")) return(.finish(oo)) else {
+      # construct an informative error... eventually there will be more 
+      # detailed info here
+      errorhint <- paste('\nThis error was generated by: ',
```
Comment (leeper) [inline R/arg_reconcile.R:102]: add spaces after commas on all lines below this

File: R/arg_reconcile.R (hunk 3 above)
Comment (bokov) [reply R/arg_reconcile.R:56]: Done

File: R/arg_reconcile.R (hunk 4 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done, you're right, backticks are dodgy

File: R/arg_reconcile.R (hunk 5 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done

File: R/arg_reconcile.R (hunk 6 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done

File: R/arg_reconcile.R (hunk 7 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done

File: R/arg_reconcile.R (hunk 8 above)
Comment (bokov) [reply R/arg_reconcile.R]: Not quite, but I found a one-liner using `duplicated` that does work.

File: R/arg_reconcile.R (hunk 9 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done, good catch, thanks.

File: R/arg_reconcile.R (hunk 10 above)
Comment (bokov) [reply R/arg_reconcile.R]: Oops, sorry. Done.

File: R/arg_reconcile.R (hunk 11 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done

File: R/arg_reconcile.R (hunk 12 above)
Comment (bokov) [reply R/arg_reconcile.R]: Done

**Comment by bokov** [comment]
Arrgh. It's the problem with CRAN again where the Windows binary packages lag behind the source packages and AppVeyor errors instead of doing the reasonable thing and falling back on the latest available binary package.

**Comment by bokov** [comment]
Failed on the 32bit developer version due to 60 min timeout. I re-ran the failed appveyor job unchanged and the second time it succeeded.

So, please try re-running it on your end. It should work.

**Comment by bokov** [comment]
Hi, just following up to see if there are any other issues I need to address. Thanks.

**Comment by bokov** [comment]
Note: after merging the latest changes from your `main` branch, many tests and examples started failing. I submitted a possible fix to those in #272 .

After that or another fix is implemented, I will merge it into this PR and push again.

**Pull request closed without merging**